it/src/test/java/com/google/cloud/teleport/it/spanner/SpannerResourceManagerUtilsTest.java

```diff
@@ -0,0 +1,176 @@
+    String testBaseString = "test-database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesNonLetterFirstCharWithLetter() {
+    String testBaseString = "0_test_database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual.charAt(0))
+        .isAnyOf(
+            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q',
+            'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z');
+    assertThat(actual).endsWith("_test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test_Database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingHyphen() {
+    String testBaseString = "test_database---";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingUnderscore() {
+    String testBaseString = "test_database___";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesDollarSignWithHyphen() {
+    String testBaseString = "test$instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).startsWith("test-instance");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesDotWithHyphen() {
+    String testBaseString = "test.instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).startsWith("test-instance");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesNonLetterFirstCharWithLetter() {
+    String testBaseString = "0-test-instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual.charAt(0))
+        .isAnyOf(
+            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q',
+            'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z');
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesUnderscoreWithHyphen() {
+    String testBaseString = "test_instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).startsWith("test-instance");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test-Instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).startsWith("test-instance");
+  }
+
+  @Test
+  public void testGenerateNewId_inputLengthLongerThanTargetLength() {
+    String longId = "long-test-id-string";
+
+    String actual = generateNewId(longId, 13);
+
+    assertThat(actual.length()).isAtMost(13);
```

Comment:
This should also use a regex. We know that the value is the right length, but what if it has invalid characters?

Reply:
Please also remove the check on the length and make it part of the regex. Regexes support an exact count (`{N}`) or a range (`{M,N}`), and since the prefix ("long-") is fixed, we know how long the rest of it should be.